staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature_support_checker.go

```diff
@@ -0,0 +1,164 @@
+		Err:       err,
+	}
+
+	return supported, err
+}
+
+// checkFeatureSupport checks the support of a specific feature.
+func (c *DefaultFeatureSupportChecker) checkFeatureSupport(ctx context.Context, feature storage.Feature) (bool, error) {
+	if feature == RequestWatchProgress {
+		return supportsRequestWatchProgress(ctx, c.client)
+	}
+	// Unimplemented features are assumed to be supported
+	return true, nil
+}
+
+// supportsRequestWatchProgressClientInterface defines the interface required for testing purposes in supportsRequestWatchProgress function of the interface.
+// It is intended to be used exclusively within the supportsRequestWatchProgress function for interacting with etcd client components.
+type supportsRequestWatchProgressClientInterface interface {
+	// Endpoints returns the list of endpoints for the etcd client.
+	Endpoints() []string
+
+	// Status retrieves the status information from the etcd client connected to the specified endpoint.
+	// It takes a context.Context parameter for cancellation or timeout control.
+	// It returns a clientv3.StatusResponse containing the status information or an error if the operation fails.
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+}
+
+// supportsRequestWatchProgress evaluates whether RequestWatchProgress supported by current version of etcd.
+// Based on this issues:
+//   - Issue #15220: https://github.com/etcd-io/etcd/issues/15220
+//   - KEP 2340: Consistent reads from cache: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#bug-in-etcd-progress-notification
+//
+// It returns a boolean indicating whether the version is deprecated and an error if any occurred during the version check.
+func supportsRequestWatchProgress(ctx context.Context, c supportsRequestWatchProgressClientInterface) (bool, error) {
+	return checkSupportsRequestWatchProgress(ctx, c)
+}
+
+func checkSupportsRequestWatchProgress(ctx context.Context, c supportsRequestWatchProgressClientInterface) (bool, error) {
+	eps := c.Endpoints()
+	g, ctx := errgroup.WithContext(ctx)
+
+	for _, ep := range eps {
+		// if cluster is current, any endpoint gives a recent version
+		// Create a new variable local to each iteration, to pass vet check
+		ep := ep
+		g.Go(func() error {
+			return checkEndpoint(ctx, c, ep)
+		})
+	}
+	err := g.Wait()
+
+	// check if the error is from type ErrNoRequestWatchProgressSupport
+	if errors.Is(err, ErrNoRequestWatchProgressSupport) {
+		return true, nil
+	}
+	return false, err
+}
+
+func checkEndpoint(ctx context.Context, c supportsRequestWatchProgressClientInterface, endpoint string) error {
+	resp, err := c.Status(ctx, endpoint)
+	if err != nil {
+		return fmt.Errorf("failed checking etcd version, endpoint: %q: %w", endpoint, err)
+	}
+	canSupport, err := checkVersion(resp.Version)
+	if err != nil {
+		return fmt.Errorf("failed parsing etcd version: %w", err)
+	}
+	if canSupport {
```

Comment:
This seems like a bug? We should return an error if it cannot support, right?

<edit> - it seems that the naming of functions/variables is just super confusing...
We should change it..